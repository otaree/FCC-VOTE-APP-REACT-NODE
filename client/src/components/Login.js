import React from 'react';
import { connect } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import axios from '../axios-poll';
import FaEnvelope from 'react-icons/lib/fa/envelope';

import * as authActions from '../store/actions/auth';
import "./auth.css";

export class Login extends React.Component {
    state = {
        email: {
            value: "",
            error: false
        },
        password: {
            value: "",
            error: false
        }
    };

    emailChangeHandler = e => {
        this.updateFieldValue("email", e.target.value);
    };

    passwordChangeHandler = e => {
        this.updateFieldValue("password", e.target.value);
    };

    submitHandler = async e => {
        e.preventDefault();

        if (this.validator()) return;

        const user = { 
            email: this.state.email.value, 
            password: this.state.password.value  
        };

        try {
            const response = await axios.post("/user/login", user);
            this.props.login({ ...response.data, token: response.headers['x-auth'] });
            this.props.history.push('/');
        } catch (e) {
            this.props.loginFail();
        }

    }

    validator = () => {
        let error = false;

        if (!isEmail(this.state.email.value)) {
            error = true;
            
            this.setState(prevState => {
                return {
                    email: {
                        value: prevState.email.value,
                        error: true
                    }
                };
            });
        }

        if (this.state.password.value.length < 6) {
            error = true;
            
            this.setState(prevState => {
                return {
                    password: {
                        value: prevState.password.value,
                        error: true
                    }
                };
            });
        }
        

       return error;
    };

    updateFieldValue = (field, value) => {
        this.setState(prevState => {
            return {
                [field]: {
                    value,
                    error: false
                }
            };
        });
    };

    render() {
        return (
            <section className="section">
                {
                    this.props.authError.isError &&
                    <div className="notification is-danger auth-error">
                        <button className="delete"></button>
                        {this.props.authError.value}
                    </div>
                }
                <div className="level">
                        <div className="level-item">
                            <div className="title">Login</div>
                        </div>
                </div>
                <div className="columns is-mobile">
                    <div className="column is-half is-offset-one-quarter">
                        <form onSubmit={this.submitHandler}>
                            <div className="field">
                                <label className="label">Email</label>
                                <div className="control has-icons-left">
                                     <input 
                                         className="input"
                                          type="email"
                                          placeholder="Email"
                                          value={this.state.email.value}
                                          onChange={this.emailChangeHandler}
                                      />
                                      <span className="icon is-small is-left">
                                         <FaEnvelope />
                                     </span>
                                      {
                                          this.state.email.error &&
                                          <div className="auth-error">Invalid email</div>
                                      }
                                </div>
                            </div>
                                  
                            <div className="field">
                                <label className="label">Password</label>
                                <div className="control">
                                     <input 
                                         className="input"
                                          type="password"
                                          placeholder="Password"
                                          value={this.state.password.value}
                                          onChange={this.passwordChangeHandler}
                                      />
                                      {
                                          this.state.password.error &&
                                          <div className="auth-error">Invalid password</div>
                                      }
                                </div>
                            </div>
                                  
                           <div className="field">
                                <div className="control">
                                     <input 
                                         className="button is-primary"
                                          type="submit"
                                          value="Login"
                                      />
                                </div>
                           </div>
                        </form>
                    </div>
                </div>
            </section>
        );
    }
}

const mapStateToProps = state => {
    return {
        authError: state.auth.error,
        loading: state.auth.loading
    };
};

const mapDispatchProps = dispatch => {
    return {
        login: (user) => dispatch(authActions.login(user)),
        loginFail: () => dispatch(authActions.fail("Invalid email/password"))
    };
};

export default connect(mapStateToProps, mapDispatchProps)(Login);