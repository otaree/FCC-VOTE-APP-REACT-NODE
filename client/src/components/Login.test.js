import React from 'react';
import { shallow, mount } from 'enzyme';


import { Login } from './Login';

describe('SignUp', () => {
    let mock;
    const mockLogin = jest.fn();
    const mockLoginFail = jest.fn();
    const mockPush = jest.fn();

    let props = {
        signUp: mockLogin,
        signUpFail: mockLoginFail,
        history: {
            push: mockPush
        },
        authError: {
            isError: false,
            value: ''
        }
    }
    let login = shallow(<Login {...props} />);

    it('should renders correctly', () => {
        expect(login.find('.auth-server-fail').length).toEqual(0);        
        expect(login).toMatchSnapshot();
    });

    describe('When from fields value changes', () => {
        beforeEach(() => {
            login = shallow(<Login {...props} />);
        });

        it('should set email on input change', () => {
            const value = "test@test.com";
            login.find('input').at(0).simulate('change', {
                target: { value }
            });
            expect(login.state('email').value).toEqual(value);
        });

        it('should set password on input change', () => {
            const value = "password!";
            login.find('input').at(1).simulate('change', {
                target: { value }
            });
            expect(login.state('password').value).toEqual(value);
        });
    });

    describe('when authentication Error', () => {
        beforeEach(() => {
            props.authError.isError = true;
            props.authError.value = "Authentication Error";
            login = shallow(<Login {...props} />);
        });

        it('should render the error message', () => {
            expect(login.find('.auth-error').length).toEqual(1);
        });
    });
});