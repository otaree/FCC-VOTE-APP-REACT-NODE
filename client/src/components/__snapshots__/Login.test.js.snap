// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    authError={
        Object {
            "isError": false,
            "value": "",
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    signUp={[Function]}
    signUpFail={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <form
          className="form"
          onSubmit={[Function]}
>
          <div
                    className="form-control"
          >
                    <input
                              onChange={[Function]}
                              placeholder="Email"
                              type="email"
                              value=""
                    />
          </div>
          <div
                    className="form-control"
          >
                    <input
                              className="input"
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                    />
          </div>
          <div
                    className="form-control"
          >
                    <input
                              className="btn"
                              type="submit"
                              value="Login"
                    />
          </div>
</form>,
      ],
      "className": "auth",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-control"
>
              <input
                            onChange={[Function]}
                            placeholder="Email"
                            type="email"
                            value=""
              />
</div>,
            <div
              className="form-control"
>
              <input
                            className="input"
                            onChange={[Function]}
                            placeholder="Password"
                            type="password"
                            value=""
              />
</div>,
            <div
              className="form-control"
>
              <input
                            className="btn"
                            type="submit"
                            value="Login"
              />
</div>,
          ],
          "className": "form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Email"
                  type="email"
                  value=""
/>,
                false,
              ],
              "className": "form-control",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Email",
                  "type": "email",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  className="input"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
/>,
                false,
              ],
              "className": "form-control",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "input",
                  "onChange": [Function],
                  "placeholder": "Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="btn"
                type="submit"
                value="Login"
/>,
              "className": "form-control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "btn",
                "type": "submit",
                "value": "Login",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <form
            className="form"
            onSubmit={[Function]}
>
            <div
                        className="form-control"
            >
                        <input
                                    onChange={[Function]}
                                    placeholder="Email"
                                    type="email"
                                    value=""
                        />
            </div>
            <div
                        className="form-control"
            >
                        <input
                                    className="input"
                                    onChange={[Function]}
                                    placeholder="Password"
                                    type="password"
                                    value=""
                        />
            </div>
            <div
                        className="form-control"
            >
                        <input
                                    className="btn"
                                    type="submit"
                                    value="Login"
                        />
            </div>
</form>,
        ],
        "className": "auth",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form-control"
>
                <input
                                onChange={[Function]}
                                placeholder="Email"
                                type="email"
                                value=""
                />
</div>,
              <div
                className="form-control"
>
                <input
                                className="input"
                                onChange={[Function]}
                                placeholder="Password"
                                type="password"
                                value=""
                />
</div>,
              <div
                className="form-control"
>
                <input
                                className="btn"
                                type="submit"
                                value="Login"
                />
</div>,
            ],
            "className": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
/>,
                  false,
                ],
                "className": "form-control",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Email",
                    "type": "email",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    className="input"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
/>,
                  false,
                ],
                "className": "form-control",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "input",
                    "onChange": [Function],
                    "placeholder": "Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="btn"
                  type="submit"
                  value="Login"
/>,
                "className": "form-control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "btn",
                  "type": "submit",
                  "value": "Login",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
